client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java

```diff
@@ -447,367 +150,431 @@ public URI getUri()
+    }
+
+    private static Properties extractPropertiesFromUri(URI uri, List<PropertyName> restrictedProperties)
     {
-        Map<String, Object> result = new HashMap<>();
+        Properties result = new Properties();
+        CatalogAndSchema catalogAndSchema = parseCatalogAndSchema(uri);
+        if (catalogAndSchema.getCatalog().isPresent() && restrictedProperties.contains(CATALOG.getPropertyName())) {
+            throw new RestrictedPropertyException(PropertyName.CATALOG, "Catalog cannot be set in the URL");
+        }
+
+        if (catalogAndSchema.getSchema().isPresent() && restrictedProperties.contains(SCHEMA.getPropertyName())) {
+            throw new RestrictedPropertyException(PropertyName.SCHEMA, "Schema cannot be set in the URL");
+        }
+
+        catalogAndSchema.getCatalog().ifPresent(value -> result.put(CATALOG.getKey(), value));
+        catalogAndSchema.getSchema().ifPresent(value -> result.put(SCHEMA.getKey(), value));
+
+        if (isSecureConnection(uri)) {
+            result.put(SSL.getKey(), SSL.encodeValue(true));
+        }
 
-        if (query == null) {
+        if (isNullOrEmpty(uri.getQuery())) {
             return result;
         }
 
-        Iterable<String> queryArgs = QUERY_SPLITTER.split(query);
-        for (String queryArg : queryArgs) {
-            List<String> parts = ARG_SPLITTER.splitToList(queryArg);
-            if (parts.size() != 2) {
-                throw new SQLException(format("Connection argument is not a valid connection property: '%s'", queryArg));
+        try {
+            Map<String, String> params = QUERY_SPLITTER.split(uri.getQuery());
+            for (Map.Entry<String, String> entry : params.entrySet()) {
+                PropertyName name = PropertyName.findByKey(entry.getKey()).orElseThrow(() -> new RuntimeException(format("Unrecognized connection property '%s'", entry.getKey())));
+                if (restrictedProperties.contains(name)) {
+                    throw new RestrictedPropertyException(name, format("Connection property %s cannot be set in the URL", name));
+                }
+                result.put(entry.getKey(), entry.getValue());
             }
 
-            String key = parts.get(0);
-            PropertyName name = PropertyName.findByKey(key).orElseThrow(() -> new SQLException(format("Unrecognized connection property '%s'", key)));
-            if (restrictedProperties.contains(name)) {
-                throw new RestrictedPropertyException(name, format("Connection property %s cannot be set in the URL", parts.get(0)));
+            return result;
+        }
+        catch (IllegalArgumentException e) {
+            String message = e.getMessage();
+            if (message.matches("Duplicate key \\[.*\\] found.")) {
+                throw new RuntimeException(format("Connection property %s is in the URL multiple times", extractPropertyName(message)));
```

Comment:
Can we catch this around the call to `put()` so we don't have to parse the error?